src/custom/state/appData/atoms.ts

```diff
@@ -0,0 +1,96 @@
+import { atom } from 'jotai'
+import { atomWithStorage } from 'jotai/utils'
+
+import {
+  AddAppDataToUploadQueueParams,
+  AppDataPendingToUpload,
+  AppDataInfo,
+  FlattenedAppDataFromUploadQueue,
+  RemoveAppDataFromUploadQueueParams,
+  UpdateAppDataOnUploadQueueParams,
+} from 'state/appData/types'
+import { buildAppDataRecordKey, parseAppDataRecordKey } from 'state/appData/utils'
+
+const INITIAL_APP_DATA_UPLOAD_QUEUE: AppDataPendingToUpload = {}
+
+/**
+ * Base atom that store the current appDataInfo
+ */
+export const appDataInfoAtom = atom<AppDataInfo | null>(null)
+
+/**
+ * Base atom that stores all appData pending to be uploaded
+ */
+export const appDataUploadQueueAtom = atomWithStorage(
+  'appDataUploadQueue', // local storage key
+  INITIAL_APP_DATA_UPLOAD_QUEUE
+)
+
+/**
+ * Write only atom to add an appData to upload queue
+ */
+export const addAppDataToUploadQueueAtom = atom(
+  null,
+  (get, set, { chainId, orderId, appData }: AddAppDataToUploadQueueParams) => {
+    set(appDataUploadQueueAtom, () => {
+      const docs = get(appDataUploadQueueAtom)
+      const key = buildAppDataRecordKey({ chainId, orderId })
+
+      return {
+        ...docs,
+        [key]: { ...appData, uploading: false },
```

Comment:
Why order id is not enough? 

Also, are we sure we want it to be an object and not an array (since it is really a queue)

Reply:
> It is, but I added it for improved debugging. Besides local debugging, I want to include this info in sentry error logs, and maybe in analytics too?

Okay, then fine with me :) 
I asked cause order id has already the chain. But debuggability is good reason

> I prefer object for O(1) access, as we need to update/delete items individually An alternative would be to update the whole list at once, but that would mean waiting all promises completing. Which is not a big deal, I just prefer this way.

I don't feel strongly. Yes, true about O(1), although in practice this will be the case for an array too. We don't grow linearly with users, or anything. The max i expect to have in this array is 1 element, at most 2, So u can see it as just a loop over 1-2 elements TOPs. Not real performance improvement

Anyways, im fine with the object form

Reply:
Refactored on https://github.com/cowprotocol/cowswap/pull/747